RecoParticleFlow/PFRecHitProducer/plugins/alpaka/PFRecHitTopologyESProducer.cc

```diff
@@ -0,0 +1,215 @@
+#include "Geometry/CaloTopology/interface/EcalBarrelTopology.h"
+#include "Geometry/CaloTopology/interface/EcalEndcapTopology.h"
+#include "RecoParticleFlow/PFRecHitProducer/interface/alpaka/CalorimeterDefinitions.h"
+
+#include <memory>
+#include <variant>
+
+namespace ALPAKA_ACCELERATOR_NAMESPACE {
+  using namespace ParticleFlowRecHitProducerAlpaka;
+
+  template <typename CAL>
+  class PFRecHitTopologyESProducer : public ESProducer {
+  public:
+    PFRecHitTopologyESProducer(edm::ParameterSet const& iConfig) : ESProducer(iConfig) {
+      auto cc = setWhatProduced(this);
+      geomToken_ = cc.consumes();
+      if constexpr (std::is_same_v<CAL, HCAL>)
+        hcalToken_ = cc.consumes();
+    }
+
+    static void fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+      edm::ParameterSetDescription desc;
+      desc.add<std::string>("appendToDataLabel", "");
+      descriptions.addWithDefaultLabel(desc);
+    }
+
+    std::unique_ptr<typename CAL::TopologyTypeHost> produce(const typename CAL::TopologyRecordType& iRecord) {
+      const auto& geom = iRecord.get(geomToken_);
+      auto product = std::make_unique<typename CAL::TopologyTypeHost>(CAL::SIZE, cms::alpakatools::host());
+      auto view = product->view();
+
+      const int calEnums[2] = {CAL::SubdetectorBarrelId, CAL::SubdetectorEndcapId};
+      for (const auto subdet : calEnums) {
+        // Construct topology
+        //  for HCAL: using dedicated record
+        //  for ECAL: from CaloGeometry (separate for barrel and endcap)
+        const CaloSubdetectorGeometry* geo = geom.getSubdetectorGeometry(CAL::DetectorId, subdet);
+        const CaloSubdetectorTopology* topo;
+        std::variant<EcalBarrelTopology, EcalEndcapTopology> topoVar;  // need to store ECAL topology temporarily
+        if constexpr (std::is_same_v<CAL, HCAL>)
+          topo = &iRecord.get(hcalToken_);
+        else if (subdet == EcalSubdetector::EcalBarrel)
+          topo = &topoVar.emplace<EcalBarrelTopology>(geom);
+        else
+          topo = &topoVar.emplace<EcalEndcapTopology>(geom);
+
+        // Fill product
+        for (auto const detId : geom.getValidDetIds(CAL::DetectorId, subdet)) {
+          const uint32_t denseId = CAL::detId2denseId(detId);
+          assert(denseId < CAL::SIZE);
+
+          const GlobalPoint pos = geo->getGeometry(detId)->getPosition();
+          view.positionX(denseId) = pos.x();
+          view.positionY(denseId) = pos.y();
+          view.positionZ(denseId) = pos.z();
+
+          for (uint32_t n = 0; n < 8; n++) {
+            uint32_t neighDetId = getNeighbourDetId(detId, n, *topo);
+            if (CAL::detIdInRange(neighDetId))
+              view.neighbours(denseId)(n) = CAL::detId2denseId(neighDetId);
+            else
+              view.neighbours(denseId)(n) = 0xffffffff;
+          }
+        }
+      }
+
+      // Remove neighbours that are not backward compatible (only for HCAL)
+      if (std::is_same_v<CAL, HCAL>)
```

Comment:
```suggestion
      if constexpr (std::is_same_v<CAL, HCAL>) {
```
?
(I'd suggest to add the curly braces to explicitly denote the code blocks for clarity)